File: 1653-number-of-good-leaf-nodes-pairs/number-of-good-leaf-nodes-pairs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        ans = 0
        def bfs(node):
            nonlocal ans
            if not node.left and not node.right:
                return [1]
            left = bfs(node.left) if node.left else []
            right = bfs(node.right) if node.right else []
            for num1 in left:
                for num2 in right:
                    if num1 + num2 <= distance:
                        ans += 1
            tot = left + right
            return [num + 1 for num in tot]   
        bfs(root)
        return ans        

File: 1653-number-of-good-leaf-nodes-pairs/number-of-good-leaf-nodes-pairs.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countPairs(root *TreeNode, distance int) int {
    ans := 0

    var dfs func(cur *TreeNode) []int
    dfs = func(cur *TreeNode) []int {
        if cur.Left == nil && cur.Right == nil {
            return []int{1}
        }

        var left, right []int
        if cur.Left != nil {
            left = dfs(cur.Left)
        } else {
            left = make([]int, 0)
        }

        if cur.Right != nil {
            right = dfs(cur.Right)
        } else {
            right = make([]int, 0)
        }

        for _, v1 := range left {
            for _, v2 := range right {
                if v1 + v2 <= distance {
                    ans++
                }
            }
        }

        left = append(left, right...)
        for i := range left {
            left[i]++
        }
        return left
    }

    dfs(root)
    return ans
}
